File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/03-RecursiveAlgorithms/PythonEx/SuperDigitTest.py
# Computes the super-digit of a number
def superDigit(n: int)->int:
    # Base case
    if n < 10: return n

    digitSum = 0
    while n > 0:
        digitSum += n % 10
        n = n // 10

    return superDigit(digitSum)

# main function
def main():
    sd = superDigit(9875)
    print(f"SuperDigit of 9875 is {sd}")

    sd = superDigit(12478453)
    print(f"SuperDigit of 12478453 is {sd}")

# Run main
if __name__ == "__main__":
    main()

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/03-RecursiveAlgorithms/JavaEx/SuperDigitTest.java
///-------------------------------------------------------------------
/**************************************************************
* Pseudocode Template for Divide & Conquer Algorithms
*
* How to solve a problem P recursively:
*
Solve(P) {
    // Base case(s)
    if P is a base case problem
        return the solution immediately

    // Divide P into P1, P2, ..Pn each of smaller scale (n>=2)
    // Solve subproblems recursively
    S1 = Solve(P1);  // Solve P1 recursively to obtain S1
    S2 = Solve(P2);  // Solve P2 recursively to obtain S2
    ...
    Sn = Solve(Pn);  // Solve Pn recursively to obtain Sn

    // Merge the solutions to subproblems
    // to get the solution to the original big problem
    S = Merge(S1, S2, ..., Sn);

    // Return the solution
    return S;
} //end-Solve
*/

public class SuperDigitTest {
    /**
     * Computes the super-digit of a number
     */
    static int superDigit(int n) {
        /* Base case */
        if (n < 10) return n;

        int digitSum = 0;
        while (n > 0) {
            digitSum += n % 10;
            n = n / 10;
        } //end-while

        return superDigit(digitSum);        
    } /* end-superDigit */

    public static void main(String[] args){
        int sd = superDigit(9875);
        System.out.printf("SuperDigit of 9875 is %d\n", sd);

        sd = superDigit(12478453);
        System.out.printf("SuperDigit of 12478453 is %d\n", sd);     
    } // end-main
} // end-SuperDigitTest
